assembly/index.ts

```diff
@@ -1,4 +1,221 @@
-import { JSONDecoder } from './decoder';
+import { JSONDecoder, JSONHandler } from './decoder';
 import { JSONEncoder } from './encoder';
 
+export namespace JSON {
+    export enum Val_Type {
+        STRING = 0,
+        NUMBER = 1,
+        BOOL = 2,
+        NULL = 3,
+        ARRAY = 4,
+        OBJECT = 5
+    }
+
+    export abstract class Value {
+        constructor(public type: Val_Type) { }
+
+        static String(str: string): Value {
+            return new Str(str)
+        }
+        static Number(num: i64): Value {
+            return new Number(num)
+        }
+        static Bool(b: bool): Value {
+            return new Bool(b)
+        }
+        static Null(): Value {
+            return new Null();
+        }
+        static Array(): Value {
+            return new Arr();
+        }
+        static Object(): Value {
+            return new Object();
+        }
+
+        get isString(): bool {
+            return this.type == Val_Type.STRING;
+        }
+
+        get isObject(): bool {
+            return this.type == Val_Type.OBJECT;
+        }
+
+        get isArray(): bool {
+            return this.type == Val_Type.ARRAY;
+        }
+
+        get isNumber(): bool {
+            return this.type == Val_Type.NUMBER;
+        }
+
+        get isBool(): bool {
+            return this.type == Val_Type.BOOL;
+        }
+
+        get isNull(): bool {
+            return this.type == Val_Type.NULL;
+        }
+    }
+
+    export class Str extends Value {
```

Comment:
given that there is no dynamic dispatch yet in AssemblyScript, do we really need these types? Why not just have `data: usize` field in `Value` with accessors for different types.
This will also allow to replace `(this.obj.get(key) as Str).str` with `this.obj.get(key).str`, etc.

Reply:
That's how I initially had it, but since null didn't have any data and number is a i64, it thought I'd go this route.  Your suggestion is good though so I'll move this logic to the Value class and make them properties.

Reply:
If you want to avoid unnecessary memory usage, then most efficient solution likely would be using tagger pointers (see https://www.mikeash.com/pyblog/friday-qa-2012-07-27-lets-build-tagged-pointers.html), but this is something which makes more sense as part of language runtime.